pkg/kubelet/checkpoint/checkpoint.go

```diff
@@ -103,7 +103,11 @@ func LoadPods(cpm checkpointmanager.CheckpointManager) ([]*v1.Pod, error) {
 			klog.Errorf("Failed to retrieve checkpoint for pod %q: %v", key, err)
 			continue
 		}
-		pods = append(pods, checkpoint.GetPod())
+		// clear NodeSelector when restoring from checkpoint otherwise pod cannot be restarted
+		// when there is no connection to api-server
+		var p *v1.Pod = checkpoint.GetPod()
+		p.Spec.NodeSelector = nil
```

Comment:
presumably the same applies to affinity rules as well (since those make use of node labels)?

does this modified version of the pod get replaced by the API version once we have a connection to the API server?

do static pods go through this path?

Reply:
Not sure whether this applies to affinity rules as well.

The pod does not get replaced once we have a connection to the API server (but does get updated if a change had been made on the server-side)

I do not believe that static pods go through this path. They would be restarted by reloading the YAML file that specified them.

Reply:
> The pod does not get replaced once we have a connection to the API server (but does get updated if a change had been made on the server-side)

Is it possible for a pod with a mismatched nodeSelector to get captured in a checkpoint if the kubelet shuts down at an inopportune time, and run when the kubelet starts back up?

Reply:
If a label is changed such that a particular pod should no longer be scheduled on a node then the kubelet should stop the pod and remove the checkpoint information. If kubelet is prevented from removing the checkpoint information (by a powerdown for example) then upon restarting that pod will be restarted from the checkpoint. As soon as the connection to the API server is restored, that pod would then be stopped.

So there is a dependency on the order in which kubelet does things (does it remove the checkpoint and then stop the pod or vice-versa?).

Reply:
I just want to make sure we understand what happens in this scenario:

* create a node "mynode" with no extra labels
* create a pod "mypod" with nodeSelector={"a":"true"} and nodeName="mynode"

is it possible for "mypod" to end up in a checkpoint for that node, even temporarily, given that its nodeSelector does not match the node's labels? is a pod added to the node's checkpoint before or after passing the admission checks in the kubelet?

Reply:
The checkpoint information for a particular pod is created when the pod starts executing on that node. So no admission => pod is not run => no checkpoint file for that pod.